libs/autotagging-commons/src/main/java/org/opensearch/rule/service/RulePersistenceService.java

```diff
@@ -0,0 +1,215 @@
+        BoolQueryBuilder boolQuery = QueryBuilders.boolQuery();
+        if (id != null) {
+            return boolQuery.must(QueryBuilders.termQuery(_ID_STRING, id));
+        }
+        for (Map.Entry<Attribute, Set<String>> entry : attributeFilters.entrySet()) {
+            Attribute attribute = entry.getKey();
+            Set<String> values = entry.getValue();
+            if (values != null && !values.isEmpty()) {
+                BoolQueryBuilder attributeQuery = QueryBuilders.boolQuery();
+                for (String value : values) {
+                    attributeQuery.should(QueryBuilders.matchQuery(attribute.getName(), value));
+                }
+                boolQuery.must(attributeQuery);
+            }
+        }
+        boolQuery.filter(QueryBuilders.existsQuery(retrieveFeatureTypeInstance().getName()));
+        return boolQuery;
+    }
+
+    /**
+     * Process searchResponse from index and send a GetRuleResponse
+     * @param searchResponse - Response received from index
+     * @param listener - ActionListener for GetRuleResponse
+     */
+    void handleGetRuleResponse(String id, SearchResponse searchResponse, ActionListener<? extends GetRuleResponse> listener) {
+        List<SearchHit> hits = Arrays.asList(searchResponse.getHits().getHits());
+        if (id != null && hits.isEmpty()) {
+            logger.error("Rule with ID " + id + " not found.");
+            listener.onFailure(new ResourceNotFoundException("Rule with ID " + id + " doesn't exist in the .rules index."));
+            return;
+        }
+        Map<String, Rule> ruleMap = hits.stream()
+            .map(hit -> parseRule(hit.getId(), hit.getSourceAsString()))
+            .filter(Objects::nonNull)
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+        String nextSearchAfter = hits.isEmpty() ? null : hits.get(hits.size() - 1).getId();
+        listener.onResponse(buildGetRuleResponse(ruleMap, nextSearchAfter, RestStatus.OK));
+    }
+
+    /**
+     * Parses a source string into a Rule object
+     * @param id - document id for the Rule object
+     * @param source - The raw source string representing the rule to be parsed
+     */
+    private Map.Entry<String, Rule> parseRule(String id, String source) {
+        try (
+            XContentParser parser = MediaTypeRegistry.JSON.xContent()
+                .createParser(NamedXContentRegistry.EMPTY, DeprecationHandler.THROW_UNSUPPORTED_OPERATION, source)
+        ) {
+            return Map.entry(id, Builder.fromXContent(parser, retrieveFeatureTypeInstance()).build());
+        } catch (IOException e) {
+            logger.info("Issue met when parsing rule for ID {}: {}", id, e.getMessage());
+            return null;
+        }
+    }
+
+    private boolean isExistingQueryGroup(String queryGroupId) {
+        return clusterService.state().metadata().queryGroups().containsKey(queryGroupId);
+    }
+
```

Comment:
This snippet of code doesn't seem to sit in the right place as it doesn't concern the persistence at all. We should create a contract (interface) for this logic which features can implement and use as per their use cases and storage layer.